src/loader/key-loader.ts

```diff
@@ -0,0 +1,137 @@
+
+interface OnKeyLoadingPayload {
+  frag: Fragment
+}
+
+interface KeyLoaderContext extends LoaderContext {
+  frag: Fragment
+}
+
+class KeyLoader extends EventHandler {
+  public loaders = {};
+  public decryptkey: Uint8Array | null = null;
+  public decrypturl: string | null = null;
+
+  constructor (hls: Hls) {
+    super(hls, Event.KEY_LOADING);
+  }
+
+  destroy (): void {
+    for (const loaderName in this.loaders) {
+      let loader = this.loaders[loaderName];
+      if (loader) {
+        loader.destroy();
+      }
+    }
+    this.loaders = {};
+
+    super.destroy();
+  }
+
+  onKeyLoading (data: OnKeyLoadingPayload) {
+    const { frag } = data;
+    const type = frag.type;
+    const loader = this.loaders[type];
+    if (!frag.decryptdata) {
+      logger.warn('Missing decryption data on fragment in onKeyLoading');
+      return;
+    }
+
+    const uri = frag.decryptdata.uri;
+    if (uri === this.decrypturl || this.decryptkey) {
+      frag.decryptdata.key = this.decryptkey;
+      this.hls.trigger(Event.KEY_LOADED, { frag: frag });
+      return;
+    }
+
+    // if uri is different from previous one or if decrypt key not retrieved yet
+    if (uri !== this.decrypturl || this.decryptkey === null) {
+      let config = this.hls.config;
+      if (loader) {
+        logger.warn(`abort previous key loader for type:${type}`);
+        loader.abort();
+      }
+      if (!uri) {
+        logger.warn('key uri is falsy');
+        return;
+      }
+
+      frag.loader = this.loaders[type] = new config.loader(config);
+      this.decrypturl = uri;
+      this.decryptkey = null;
+
+      let loaderContext: KeyLoaderContext = {
+        url: uri,
+        frag: frag,
+        responseType: 'arraybuffer'
+      };
+
+      // maxRetry is 0 so that instead of retrying the same key on the same variant multiple times,
+      // key-loader will trigger an error and rely on stream-controller to handle retry logic.
+      // this will also align retry logic with fragment-loader
+      let loaderConfig: LoaderConfiguration = {
+        timeout: config.fragLoadingTimeOut,
+        maxRetry: 0,
+        retryDelay: config.fragLoadingRetryDelay,
+        maxRetryDelay: config.fragLoadingMaxRetryTimeout
+      };
+
+      let loaderCallbacks: LoaderCallbacks<KeyLoaderContext> = {
+        onSuccess: this.loadsuccess.bind(this),
+        onError: this.loaderror.bind(this),
+        onTimeout: this.loadtimeout.bind(this)
+      };
+
+      frag.loader.load(loaderContext, loaderConfig, loaderCallbacks);
+    }
+  }
+
+  loadsuccess (response: LoaderResponse, stats: LoaderStats, context: KeyLoaderContext) {
+    let frag = context.frag;
+    if (!frag.decryptdata) {
+      logger.error('after key load, decryptdata unset');
+      return;
+    }
+    this.decryptkey = frag.decryptdata.key = new Uint8Array(response.data as ArrayBuffer);
+
+    // detach fragment loader on load success
+    frag.loader = undefined;
+    this.loaders[frag.type] = undefined;
```

Comment:
wonder if there's a reason we don't
```suggestion
    delete this.loaders[frag.type];
```
? 🤔

Reply:
Yeah I think that makes sense. The earlier `for..in` loops then won't get undefined keys. Might make it possible to drop the conditional check inside the loop eventually. I'm down.